classes/controller/FrontController.php

```diff
@@ -477,12 +477,63 @@ public function init()
             $this->displayRestrictedCountryPage();
         }
 
+        if (Tools::isSubmit('live_edit') && !$this->checkLiveEditAccess()) {
+            Tools::redirect('index.php?controller=404');
+        }
+
         $this->iso = $iso;
         $this->context->cart = $cart;
         $this->context->currency = $currency;
     }
 
     /**
+     * Checks if the user can use Live Edit feature
+     *
+     * @return bool
+     */
+    public function checkLiveEditAccess()
+    {
+        if (!Tools::isSubmit('live_edit') || !Tools::getValue('ad') || !Tools::getValue('liveToken')) {
+            return false;
+        }
+
+        if (Tools::getValue('liveToken') != $this->validateToken()) {
+            return false;
+        }
+
+        $adminDir = str_ireplace('.', '', Tools::getValue('ad'));
```

Comment:
You understood correctly. This is incomplete though : if stripping `.` protects from parent directory paths, it doesn't protect from subdirectory paths. After a quick research I found a clean way :
```php
$adminDir = basename(Tools::getValue('ad'));
$adminDir = str_replace('.' '', $adminDir);
if ('' === $adminDir) {
    return false;
}
```
`basename()` returns the trailing name component (without extension suffix, if any). But it might return `.` or `..`. That's why we need to strip the `.` character, then test if result is an empty string. Then we can go on with `realpath()`.

Reply:
Super,
Good job.
Thank you.